api/src/main/java/org/openmrs/module/registrationcore/api/impl/PatientCreationListener.java

```diff
@@ -0,0 +1,101 @@
+package org.openmrs.module.registrationcore.api.impl;
+
+import org.openmrs.Patient;
+import org.openmrs.PatientIdentifier;
+import org.openmrs.User;
+import org.openmrs.api.UserService;
+import org.openmrs.api.context.Context;
+import org.openmrs.event.Event;
+import org.openmrs.module.registrationcore.RegistrationCoreConstants;
+import org.openmrs.module.registrationcore.api.mpi.common.MpiProperties;
+import org.openmrs.module.registrationcore.api.mpi.openempi.OpenEmpiPatientQuery;
+
+import javax.jms.Message;
+
+public class PatientCreationListener extends PatientActionListener {
+
+    private MpiProperties mpiProperties;
+
+    private IdentifierBuilder identifierBuilder;
+
+    private UserService userService;
+
+    public void setMpiProperties(MpiProperties mpiProperties) {
+        this.mpiProperties = mpiProperties;
+    }
+
+    public void setIdentifierBuilder(IdentifierBuilder identifierBuilder) {
+        this.identifierBuilder = identifierBuilder;
+    }
+
+    public void setUserService(UserService userService) {
+        this.userService = userService;
+    }
+
+    public void init() {
+        Event.subscribe(RegistrationCoreConstants.REGISTRATION_EVENT_TOPIC_NAME, this);
+    }
+
+    @Override
+    public void performMpiAction(Message message) {
+        Patient patient = extractPatient(message);
+
+        OpenEmpiPatientQuery exportedPatientQuery = pushPatientToMpiServer(patient);
+
+        updatePatient(patient, message, exportedPatientQuery.getPersonId());
+    }
+
+    private OpenEmpiPatientQuery pushPatientToMpiServer(Patient patient) {
+        return coreProperties.getMpiProvider().exportPatient(patient);
+    }
+
+    private void updatePatient(Patient patient, Message message, Integer personId) {
+        grandPrivileges();
+        try {
+            User creator = extractPatientCreator(message);
+            addPersonIdentifier(patient, creator, personId);
+            patientService.savePatient(patient);
+        } finally {
+            resetPrivileges();
+        }
+    }
+
+    private User extractPatientCreator(Message message) {
```

Comment:
@djazayeri yep. There is a bug. "KEY_REGISTERER_UUID" Always has random generated value. So we can't get "creator" by that uuid. To fix that I use now Id.